### PR TITLE
Shuttle and Pattern models and routes

### PR DESCRIPTION
**Pull request by Scttpr**
Checklist

- [x] I've added a mention in CHANGELOG.md

### CODE REVIEW

**Comment by Scttpr** [comment]
[CRUD gestion des boucles](https://trello.com/c/FNjSrEkH/389-crud-gestion-des-boucles)

File: routes/rides.mjs [hunk 1]
```diff
@@ -137,13 +142,13 @@ const router = generateCRUD(Ride, {
...
     preMiddlewares: [
-      prefetchRideMiddleware(),
+      prefetchMiddleware(Ride),
     ],
     lean: false,
     right: [CAN_GET_RIDE, CAN_GET_OWNED_RIDE, CAN_GET_RIDE_WITH_TOKEN],
     async main(ctx) {
       const { params: { id } } = ctx;
-      const ride = getPrefetchedRide(ctx, id);
+      const ride = getPrefetchedDocument(ctx, id, RIDE_MODEL_NAME);
```
Comment (Nainterceptor) [inline routes/rides.mjs:151]: If you use Ride Model to `prefetchMiddleware` l.145, it's better to use Ride Model in this function too, it isn't ?

File: routes/shuttles.mjs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  CAN_EDIT_OWNED_RIDE,
+  CAN_DELETE_SHUTTLE,
+} from '../models/rights';
+import maskOutput from '../middlewares/mask-output';
+import { ensureThatFiltersExists } from '../middlewares/query-helper';
+import ioEmitMiddleware from '../middlewares/io-emit';
+import { getPrefetchedDocument, prefetchMiddleware } from '../helpers/prefetch-document';
+import { SHUTTLE_MODEL_NAME } from '../models/helpers/constants';
+
+// @todo write tests for these routes
```
Comment (Nainterceptor) [inline routes/shuttles.mjs:22]: 👀

File: routes/shuttles.mjs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  update: {
+    preMiddlewares: [prefetchMiddleware(Shuttle)],
+    right: [CAN_EDIT_RIDE, CAN_EDIT_OWNED_RIDE],
+    middlewares: [ioEmitMiddleware(SHUTTLE_UPDATE_EVENT, [
+      'shuttle', 'campus', 'driver',
+    ])],
+    async main(ctx) {
+      const { body } = ctx.request;
+      const { id } = ctx.params;
+      const shuttle = getPrefetchedDocument(ctx, id, SHUTTLE_MODEL_NAME);
```
Comment (Nainterceptor) [inline routes/shuttles.mjs:73]: Same thing here, not consistent with prefetch

Comment (Nainterceptor) [reply routes/shuttles.mjs:73]: + Maybe you can inject getPrefetchedDocument in ctx, via the prefetchMiddleware

File: routes/shuttles.mjs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  },
+  delete: {
+    right: CAN_DELETE_SHUTTLE,
+  },
+});
+
+/*
+* @todo specific shuttle routes :
+*   - GET position
+* */
```
Comment (Nainterceptor) [inline routes/shuttles.mjs:93]: Maybe will not be necessary in the product, to check in team meeting.

File: tests/helpers/crud/create.mjs [hunk 5]
```diff
@@ -22,7 +22,6 @@ export const testCreate = (Model, {
         .query({ ...queryParams, mask: 'id' })
         .set(...roleGenerator())
         .send(cleanObject(dummyObject));
-
```
Comment (Nainterceptor) [inline tests/helpers/crud/create.mjs:25]: 👀 When it's not required by linter, dont commit this kind of change when you're not doing anything else in the file.

(It's OK when you're editing the file elsewhere, but some bots check files changed to calculate impact of the PR)

Comment (Nainterceptor) [reply tests/helpers/crud/create.mjs:25]: (No, it's not useful to rollback the change, it was a simple comment)

File: models/shuttle.mjs [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    _id: {
+      type: String,
+      required: true,
+      alias: 'id',
+    },
+    label: {
+      type: String,
+      required: true,
+    },
+    passengers: [{
```
Comment (Nainterceptor) [inline models/shuttle.mjs:84]: imo, passengers should not be stored here.

File: app.mjs [hunk 7]
```diff
@@ -29,7 +29,7 @@ app.use(errorHandler);
 app.use(helmet());
 app.use(jwt({ secret: config.get('token:secret'), passthrough: true }));
 app.use(injectUserMayMiddleware);
-app.use(pino());
+app.use(pino({ level: 'fatal' }));
```
Comment (Nainterceptor) [inline app.mjs:32]: better to store this level in environment variable, we will need INFO in production

File: models/rules.mjs [hunk 8]
```diff
@@ -44,13 +45,21 @@ export const roleEditingRule = ruleGenerator((
 ) => campusId && !!campuses.find((c) => c._id === campusId));
 
 export const tokenRideRule = ruleGenerator((_, ctx) => {
-  const ride = getPrefetchedRide(ctx, ctx.params.id);
+  // eslint-disable-next-line no-unused-vars,prefer-const
+  let [, modelName] = ctx.url.split('/');
+  const Model = mongoose.model(modelName);
+  // modelName = modelName[0].toUpperCase() + modelName.slice(1, -1);
```
Comment (Nainterceptor) [inline models/rules.mjs:51]: why commented code ?

File: models/rules.mjs [hunk 9]
```diff
@@ -44,13 +45,21 @@ export const roleEditingRule = ruleGenerator((
 ) => campusId && !!campuses.find((c) => c._id === campusId));
 
 export const tokenRideRule = ruleGenerator((_, ctx) => {
-  const ride = getPrefetchedRide(ctx, ctx.params.id);
+  // eslint-disable-next-line no-unused-vars,prefer-const
+  let [, modelName] = ctx.url.split('/');
```
Comment (Nainterceptor) [inline models/rules.mjs:49]: You should be able to remove eslint-disable and use const here, you're not mutating modelName anywhere

File: models/rules.mjs [hunk 10]
```diff
@@ -44,13 +45,21 @@ export const roleEditingRule = ruleGenerator((
...
+  const Model = mongoose.model(modelName);
+  // modelName = modelName[0].toUpperCase() + modelName.slice(1, -1);
+  const ride = getPrefetchedDocument.call(ctx, ctx.params.id, Model);
   return ride.token === ctx.query.token;
 });
 
 export const ownedRideRule = ruleGenerator((_, ctx, entity) => {
   const { user } = ctx.state;
-  const ride = entity || getPrefetchedRide(ctx, ctx.params.id);
+  // eslint-disable-next-line no-unused-vars,prefer-const
```
Comment (Nainterceptor) [inline models/rules.mjs:58]: You should be able to remove eslint-disable and use const here, you're not mutating modelName anywhere

File: models/rules.mjs [hunk 11]
```diff
@@ -44,13 +45,21 @@ export const roleEditingRule = ruleGenerator((
...
+  let [, modelName] = ctx.url.split('/');
+  const Model = mongoose.model(modelName);
+  // modelName = modelName[0].toUpperCase() + modelName.slice(1, -1);
```
Comment (Nainterceptor) [inline models/rules.mjs:61]: why commented code ?

File: models/shuttle.mjs [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+ShuttleSchema.methods.sendSMS = async function sendUserSMS(body) {
+  await Promise.all(this.passengers.map(async ({ phone }) => {
+    try {
+      if (phone) {
+        return await sendSMS(phone, body);
+      }
+    } catch (e) {
+      // Silent error
+      // eslint-disable-next-line no-console
+      console.error('\x1b[31m', e);
```
[the remark this answers is not in the document]
Comment (Nainterceptor) [reply models/shuttle.mjs:296]: Still here, what's the purpose ?

File: routes/shuttle-factories.mjs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+import {
+  CAN_CREATE_PATTERN,
+  CAN_LIST_PATTERN,
+  CAN_GET_PATTERN,
+  CAN_UPDATE_PATTERN,
+  CAN_DELETE_PATTERN,
+} from '../models/rights';
+import maskOutput from '../middlewares/mask-output';
+
+// @todo write tests for these routes
```
Comment (Nainterceptor) [inline routes/shuttle-factories.mjs:12]: I've seen tests, maybe you can remove this todo ?

File: routes/campuses/drivers.mjs [hunk 14]
```diff
@@ -42,6 +42,9 @@ router.get(
       && (ctx.query.filters.start && ctx.query.filters.end)) {
       const start = new Date(ctx.query.filters.start);
       const end = new Date(ctx.query.filters.end);
+      if (ctx.query.filters.licences) {
+        ctx.filters.licences = ctx.query.filters.licences;
```
Comment (Nainterceptor) [inline routes/campuses/drivers.mjs:46]: If you've nothing to cast, you could use addFilter middleware

**Pull request merged**